app/questionnaire/router.py

```diff
@@ -0,0 +1,75 @@
+class Router:
+
+    def __init__(self, schema, routing_path, completeness, current_location=None):
+        self._schema = schema
+        self._routing_path = routing_path
+        self._completeness = completeness
+        self._current_location = current_location
+
+        self._first_incomplete_block_in_section = None
+        self._first_incomplete_block_in_survey = None
+        self._section = None
+        self._block = None
+
+    def can_access_location(self):
+        if self._is_valid_location() and not self._is_skipping_to_section_end_or_survey_end():
+            return True
+
+        return False
+
+    def get_next_location(self):
+        if not self._is_valid_location():
+            next_location = self._get_expected_location()
+            return next_location
+
+        if self._is_skipping_to_section_end_or_survey_end():
+            return self._first_incomplete_block_in_section or self._first_incomplete_block_in_survey
+
+        return self._get_expected_location()
+
+    def _is_valid_location(self):
+        return self._current_location in self._routing_path
+
+    def _get_expected_location(self):
+        next_location = self._routing_path[0]
+
+        if self._completeness.completed_blocks:
+            next_location = self._completeness.get_first_incomplete_location_in_survey()
+            if not next_location:
+                next_location = self._routing_path[-1]
+
+        return next_location
+
+    def _is_skipping_to_section_end(self):
+        if self._first_incomplete_block_in_section:
+            return self._first_incomplete_block_in_section
+
+        expected_location = self._completeness.get_first_incomplete_location_in_section(self._section)
+
+        if self._block['type'] == 'SectionSummary' and self._current_location != expected_location:
+            self._first_incomplete_block_in_section = expected_location
+
+        return self._first_incomplete_block_in_section
+
+    def _is_skipping_to_survey_end(self):
+        if self._first_incomplete_block_in_survey:
```

Comment:
Discussed IRL, I think we can simplify things here by removing the instance variables and just getting locally as required.